lsp/src/Language/LSP/VFS.hs

```diff
@@ -346,6 +352,52 @@ changeChars logger str start finish new = do
 
 -- ---------------------------------------------------------------------
 
+-- | A position, like a 'J.Position', but where the offsets in the line are measured in Unicode code points
+-- instead of UTF-16 code units.
+data CodePointPosition =
+  CodePointPosition
+    { -- | Line position in a document (zero-based).
+      _line      :: J.UInt
+      -- | Character offset on a line in a document in *code points* (zero-based).
+    , _character :: J.UInt
+    } deriving (Show, Read, Eq, Ord)
+
+-- | Given a virtual file, translate a 'CodePointPosition' in that file into a 'J.Position' in that file.
+--
+-- We need the file itself because this requires translating between code points and code units.
+codePointPositionToPosition :: VirtualFile -> CodePointPosition -> J.Position
+codePointPositionToPosition vFile (CodePointPosition l cpc) =
+  let text = _file_text vFile
+      (_, suffix) = Rope.splitAtLine (fromIntegral l) text
+      (utf16Line, _) = Rope.splitAtLine 1 suffix
+
+      utfLine = URope.fromText $ Rope.toText utf16Line
+      -- Split at the given position in *code points*
+      (utfLinePrefix, _) = URope.splitAt (fromIntegral cpc) utfLine
+      utf16LinePrefix = Rope.fromText $ URope.toText utfLinePrefix
+      -- Get the length of the prefix in *code units*
+      cuc = Rope.length utf16LinePrefix
+  in J.Position (fromIntegral l) (fromIntegral cuc)
+
+-- | Given a virtual file, translate a 'J.Position' in that file into a 'CodePointPosition' in that file.
+-- May fail if the requested position lies inside a code point.
+--
+-- We need the file itself because this requires translating between code unit and code points.
+positionToCodePointPosition :: VirtualFile -> J.Position -> Maybe CodePointPosition
+positionToCodePointPosition vFile (J.Position l cuc) = do
+  let text = _file_text vFile
+      (_, suffix) = Rope.splitAtLine (fromIntegral l) text
+      (utf16Line, _) = Rope.splitAtLine 1 suffix
+
+  -- Split at the given position in *code units*
+  (utf16LinePrefix, _) <- Rope.splitAt (fromIntegral cuc) utf16Line
+  let utfLinePrefix = URope.fromText $ Rope.toText utf16LinePrefix
+      -- Get the length of the prefix in *code points*
+      cpc = URope.length utfLinePrefix
```

Comment:
It's a bit cleaner to avoid `URope.fromText . Rope.toText` conversions:
```suggestion
  let utfLinePrefix = Rope.toTextLines utf16LinePrefix
      -- Get the length of the prefix in *code points*
      cpc = ULines.length utfLinePrefix
```

Reply:
I guess I thought it would avoid the reliance on your unreleased change, but sure.

Reply:
Ah, right, makes sense. Conversion of a single line should not be expensive.

Reply:
If we can stick to the released API, let's not block on me making a new release of `text-rope`.

Reply:
That's the hope. No doubt this could still be cheaper, but at this point we've bounded it to logarithmic in the number of lines and linear in the length of the line, which isn't too bad.